### PR TITLE
h2, codec: Skip setting data_flags for empty trailers

### PR DESCRIPTION
**Pull request by dio**
Commit Message: This skips setting `data_flags` for empty pending trailers.

Risk Level: Low
Testing: integration test, fuzz
Docs Changes: N/A
Release Notes: N/A

Fixes: https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=24714

### CODE REVIEW

**Comment by asraa** [comment]
@dio here's a regression test:

```
+TEST_P(Http2IntegrationTest, EmptyTrailers) {
+  initialize();
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+
+  auto encoder_decoder = codec_client_->startRequest(default_request_headers_);
+  request_encoder_ = &encoder_decoder.first;
+  auto response = std::move(encoder_decoder.second);
+  codec_client_->sendData(*request_encoder_, 100000, false);
+  Http::TestRequestTrailerMapImpl request_trailers;
+  codec_client_->sendTrailers(*request_encoder_, request_trailers);
+
+  waitForNextUpstreamRequest();
+
+  upstream_request_->encodeHeaders(default_response_headers_, true);
+  response->waitForEndStream();
+  ASSERT_TRUE(response->complete());
+}


```

File: test/common/http/codec_impl_fuzz_test.cc [hunk 1]
```diff
@@ -348,9 +348,13 @@ class HttpStream : public LinkedObject<HttpStream> {
               .WillRepeatedly(InvokeWithoutArgs(
                   [&] { directionalAction(response_, stream_action.dispatching_action()); }));
         } else if (request_action == test::common::http::DirectionalAction::kTrailers) {
-          EXPECT_CALL(request_.request_decoder_, decodeTrailers_(_))
-              .WillOnce(InvokeWithoutArgs(
-                  [&] { directionalAction(response_, stream_action.dispatching_action()); }));
+          // TODO(dio): This should be controlled by runtime feature flag
+          // (envoy.reloadable_features.http2_skip_encoding_empty_trailers).
+          if (!stream_action.request().trailers().headers().empty()) {
```
Comment (dio) [inline test/common/http/codec_impl_fuzz_test.cc:353]: @asraa do you know how to honor runtime flag this test file?

File: test/common/http/codec_impl_fuzz_test.cc [hunk 2]
```diff
@@ -348,9 +348,13 @@ class HttpStream : public LinkedObject<HttpStream> {
               .WillRepeatedly(InvokeWithoutArgs(
                   [&] { directionalAction(response_, stream_action.dispatching_action()); }));
         } else if (request_action == test::common::http::DirectionalAction::kTrailers) {
-          EXPECT_CALL(request_.request_decoder_, decodeTrailers_(_))
-              .WillOnce(InvokeWithoutArgs(
-                  [&] { directionalAction(response_, stream_action.dispatching_action()); }));
+          // TODO(dio): This should be controlled by a runtime feature flag
+          // (envoy.reloadable_features.http2_skip_encoding_empty_trailers).
```
Comment (dio) [inline test/common/http/codec_impl_fuzz_test.cc:352]: @asraa do you know how to honor runtime flag within this test file?

File: source/common/http/http2/codec_impl.cc [hunk 3]
```diff
@@ -381,10 +381,15 @@ ssize_t ConnectionImpl::StreamImpl::onDataSourceRead(uint64_t length, uint32_t*
     if (local_end_stream_ && pending_send_data_.length() <= length) {
       *data_flags |= NGHTTP2_DATA_FLAG_EOF;
       if (pending_trailers_to_encode_) {
-        // We need to tell the library to not set end stream so that we can emit the trailers.
-        *data_flags |= NGHTTP2_DATA_FLAG_NO_END_STREAM;
-        submitTrailers(*pending_trailers_to_encode_);
-        pending_trailers_to_encode_.reset();
+        const bool skip_encoding_empty_trailers =
```
Comment (asraa) [inline source/common/http/http2/codec_impl.cc:384]: quick question for my own understanding: if we do skip encoding empty trailers, is it ok that they are not `reset()`? 
Just want to make sure it's not possible to trigger https://github.com/envoyproxy/envoy/blob/34fc31a98e6ecf2c8cd90efa8eeb6c56e02958d0/source/common/http/http2/codec_impl.cc#L230

Comment (dio) [reply source/common/http/http2/codec_impl.cc:384]: Good question. I’ll do some digging on the possibilities. However I think @alyssawilk could shed some light here.

Comment (dio) [reply source/common/http/http2/codec_impl.cc:384]: Yeah I we should add test on that (probably setting that null when we skip helps).

Comment (dio) [reply source/common/http/http2/codec_impl.cc:384]: I think you have pointed a valid concern. I pushed a change to make sure we do reset. Thanks, @asraa!

**Comment by dio** [comment]
> Does nghttp2 just not actually end the stream in the skip case?

Yes, when we have pending data and `pending_trailers_to_encode_` as non-nullptr we set `NGHTTP2_DATA_FLAG_NO_END_STREAM` to data_flags, add more empty data, but we never end it.

**Comment by dio** [comment]
Do retest, since: ERROR: /source/test/common/access_log/BUILD:33:14: output 'test/common/access_log/access_log_manager_impl_test/coverage.dat' was not created

**Comment by mattklein123** [comment]
> Yes, when we have pending data and pending_trailers_to_encode_ as non-nullptr we set NGHTTP2_DATA_FLAG_NO_END_STREAM to data_flags, add more empty data, but we never end it.

Yes but how does it stack overflow?

**Comment by dio** [comment]
I hope I can describe this correctly: Given the trailers map is empty and we have pending data: when we do `submitTrailers`, it actually resumes the submit data which triggers `onDataSourceRead` which calls `submitTrailers` again and again (with no end stream is set).

Thinking about this again. It seems like we should not set `pending_trailers_to_encode_` for the first place for this case.

**Comment by mattklein123** [comment]
> Thinking about this again. It seems like we should not set pending_trailers_to_encode_ for the first place for this case (2b914ec).

Yup ok this makes sense. Thank you!

**Pull request merged**